src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -999,9 +999,17 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
         case concatenations =>
           bc.genStartConcat(tree.pos)
           for (elem <- concatenations) {
-            val kind = tpeTK(elem)
-            genLoad(elem, kind)
-            bc.genStringConcat(kind, elem.pos)
+            val loadedElem = elem match {
+              case Apply(boxOp, value :: Nil) if currentRun.runDefinitions.isBox(boxOp.symbol) =>
+                // Eliminate boxing of primitive values. Boxing is introduced by erasure because
+                // there's only a single synthetic `+` method "added" to the string class.
+                value
```

Comment:
This will incorrectly transform:

```
"xx" + {println("side effect"); Int}.box(1)
```

to 

```
val sb = new StringBuilder(); sb.append("xx"); sb.append(1)
```

However, this isn't the only place in the backend that has this bug:

```
scala> {???; Int}.box(1)
res4: Integer = 1
```

We can treat this as a separate bug: https://github.com/scala/scala-dev/issues/85